web_src/js/features/repo-legacy.js

```diff
@@ -613,15 +613,15 @@ function initRepoIssueCommentEdit() {
   $(document).on('click', '.edit-content', onEditContent);
 
   // Quote reply
-  $(document).on('click', '.quote-reply', function (event) {
+  $(document).on('click', '.quote-reply', async function (event) {
+    event.preventDefault();
```

Comment:
This `preventDefault` is still ineffective I think, see https://github.com/github/eslint-plugin-github/blob/main/docs/rules/async-preventdefault.md for an explanation. I think we should move the rest of the function body into an async IIFE so that `preventDefault` is executed synchronously.

Reply:
It is effective, because there is no `await`.

You can try it by yourself https://jsfiddle.net/te9o27rg/2/

And IIRC I have explained this problem in an old issue and in the frontend guideline before.

Reply:
See this one (I cleaned the document, removed some details because I thought everyone should already know this problem)

https://github.com/wxiaoguang/gitea/blob/7de594d3544addfc4ea5e83a72ecd759943dff35/docs/content/doc/developers/guidelines-frontend.md

Reply:
Ok interesting that this works but I guess this is a common enough source of error so I guess browser vendors have made it work.

Reply:
According to https://stackoverflow.com/a/37146788/808699 it is fine to call before the first `await`.

Reply:
No, it's not "browser vendors have made it work.", that's just how JS works.

You can always rewrite the `async` function to a `Promise` style, no `async` no `await`, then everything is clear.

Reply:
> According to https://stackoverflow.com/a/37146788/808699 it is fine to call before the first `await`.

I guess my issue&PR&document are far more detailed than that SO answer ..... when I was doing the refactoring at that time, I have requested reviewing from you many times ......

Reply:
That function equality above is incorrect. Nothing executes synchronously inside an async function. See [this](https://gist.github.com/joeytwiddle/ff0a9cf92be0366b274e4283a8b5a935) for a better example.

Reply:
There is a comment: https://gist.github.com/joeytwiddle/ff0a9cf92be0366b274e4283a8b5a935?permalink_comment_id=4128782#gistcomment-4128782

> This is not exactly the truth. An async function will be executed in sync until it is converted into a Promise by reaching a return or throw statement or the execution is suspendend when a await is reached.

It just matches my example. (I didn't put error/exception in it to keep it simple).

Reply:
[MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function) also has a word on this:

> The body of an async function can be thought of as being split by zero or more await expressions. Top-level code, up to and including the first await expression (if there is one), is run synchronously. In this way, an async function without an await expression will run synchronously. If there is an await expression inside the function body, however, the async function will always complete asynchronously. 

The last sentence contradicts the first one, but I guess the first one may be what is actually happening.

Reply:
I think the last sentence's meaning is also correct.

Sentence 1: `zero or more await expressions`: every `await` handles a Promise block

Sentence 2: `up to and including the first await expression .. synchronously`: that's also correct, think about an example:

```js
function () {
    data = {}
    await fetch(url, data)
    ...
}
```

The `fetch` itself is executed synchronously and returns a Promise, `await` handles a the `fetch` returned Promise


Sentence 3: `an async function without an await expression will run synchronously`: that's simply true.

Sentence 4:  `If there is an await expression ..  always complete asynchronously`:  that's also true, because after the first `await`, the following code is already in Promise blocks.

MDN is right.

Reply:
And think about my `preventDefault` example above, you could even do https://jsfiddle.net/94awkvu7/

```js
$('#btn').on('click', async function(e) {
    await e.preventDefault(); // do not really do so. just to show what is async & await
})
```

Although it is anti-pattern, indeed it could also prevent the default successfully, because `e.preventDefault()` is really executed synchronously.

Reply:
You're right the important part is "complete asynchronously", even if part of it ran synchronously, the function will finish async when there is an await.

Reply:
Were there any other reservations here? Just wanted to ask quick since this is in the merge queue. 🙂

Reply:
IMO this PR has fixed the "quote - reply" bug and this PR also contains another important `onChange` fix, so maybe it could be merged, and then propose new fixes later.

Reply:
> Were there any other reservations here? Just wanted to ask quick since this is in the merge queue. 🙂

Not from me, the line is fine.